File: 20221109_Wednesday/20221109.go
// Leetcode: 901. Online Stock Span.
// https://leetcode.com/problems/online-stock-span/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 196 ms, faster than 77.92% of Go online submissions for Online Stock Span.
// Memory Usage: 10.4 MB, less than 6.49% of Go online submissions for Online Stock Span.
// 2022.11.09 Daily Challenge.
import "math"
type StockSpanner struct {
  i int // inedex of last element. Number of an input price.
  sv []int // stack values.
  si [] int // stack indices.
}


func Constructor() StockSpanner {
  data := StockSpanner{i:0}
  data.i = 0
  data.sv = make([]int,1,int(math.Pow10(5))) // Capacity to allocate once.
  data.si = make([]int,1,int(math.Pow10(5))) 
  data.sv[0] = int(math.Pow10(7))
  data.si[0] = 0
  return data
}


func (this *StockSpanner) Next(price int) int {
  this.i += 1
  last := len(this.sv) - 1
  for ; 0 < len(this.sv) && this.sv[last] <= price ; last -= 1 {
    this.sv = this.sv[0:last]
    this.si = this.si[0:last]
  }
  this.sv = append(this.sv,price)
  this.si = append(this.si,this.i)
  s := len(this.si)
  return  this.i - this.si[s-2]
}


/**
 * Your StockSpanner object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Next(price);
 */

File: 20221109_Wednesday/20221109.rb
# Leetcode: 901. Online Stock Span.
# https://leetcode.com/problems/online-stock-span/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 248 ms, faster than 92.31% of Ruby online submissions for Online Stock Span.
# Memory Usage: 215.3 MB, less than 15.38% of Ruby online submissions for Online Stock Span.
# 2022.11.09 Daily Challenge.
class StockSpanner
    def initialize()
      @sv = [10**7] # stack values.
      @si = [0] # stack indices.
      @i = 0
    end


=begin
    :type price: Integer
    :rtype: Integer
=end
    def next(price)
      @i += 1
      while (!@sv.empty?) && @sv.last <= price
        @sv.pop()
        @si.pop()
      end
      @sv.push(price)
      @si.push(@i)
      if 0 == @i
        return 1
      elsif 1 == @sv.size
        return @i
      else
        return @i - @si[-2]
      end
    end
end

# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner.new()
# param_1 = obj.next(price)
